R/Spatial.R

```diff
@@ -0,0 +1,180 @@
+
+#' @title Q3_Normalization
+#'
+#' @description a normalization method that uses the 75th percentile to find size factors for data normalization.
+#' @param seuratObj The Seurat object that holds the data to be normalized.
+#' @param assay The assay holding the raw data - the counts slot will be assumed to hold raw data.
+#' @return Returns a Seurat object containing normalized counts stored in the counts slot of the Q3 assay. 
+#' @export
+
+Q3_Normalization <- function(seuratObj, assay = "RNA"){
+  #from here: https://www.dlongwood.com/wp-content/uploads/2021/05/WP_MK2833_Introduction_to_GeoMx_Normalization_CTA_R5-1.pdf page 7
+  # first, divide all the genes per AOI by their respective Q3 count, 
+  #ii) second, multiply all the genes in all AOIs by a constant, defined as the geometric mean of Q3 counts for all AOIs.
+  
+  #calculate column (ROI) Q3s
+  counts <- as.matrix(seuratObj[[assay]]@counts)
+  Q3_Sample_Counts <- apply(counts,MARGIN = 2,FUN = quantile,prob=c(.75))
+  #Ensure no downstream division by zero errors
+  if(any(Q3_Sample_Counts == 0)){
+    stop("Error: 75th percentile is zero for some samples. Q3 is either too low to normalize by, or the data is too sparse.")
+  }
+  #divide count matrix by vector of Q3s
+  Q3_Divided_Counts <- counts/Q3_Sample_Counts
+  #multiply Q3 divided counts by geometric mean of Q3s (scalar)
+  Q3_Final_Counts <- Q3_Divided_Counts * compositions::geometricmean(Q3_Sample_Counts)
+  
+  #stash into seurat object under new assay
+  seuratObj[['Q3']] <- Seurat::CreateAssayObject(as(Q3_Final_Counts, "dgCMatrix"))
+  
+  return(seuratObj)
+}
+
+#' @title RUVg_Housekeeping_Normalization
+#'
+#' @description a normalization method that uses residuals from a fitted model to Remove Unwanted Variation by leveraging housekeeping genes (https://www.nature.com/articles/nbt.2931).
+#' @param seuratObj The Seurat object that holds the data to be normalized.
+#' @param assay The assay holding the raw data - the counts slot will be assumed to hold raw data.
+#' @param k the number of factors of unwanted variation. See section "Choice of number of factors of unwanted variation." in paper cited in the description.
+#' @return Returns a Seurat object containing normalized counts stored in the counts slot of the RUVg assay. 
+#' @export
+RUVg_Housekeeping_Normalization <- function(seuratObj, assay = "RNA", k = 1){
+  housekeeping_genes <- c("ABCF1", "ACTB", "ALAS1", "B2M", "CLTC", "G6PD", "GAPDH", 
+                          "GUSB", "HPRT1", "LDHA", "PGK1", "POLR1B", "POLR2A",
+                          "RPL19", "RPLP0", "SDHA", "TBP", "TUBB")
+  counts <- as.matrix(seuratObj[[assay]]@counts)
+  rounded_counts <- round(as.matrix(counts))
+  RUVg <- RUVSeq::RUVg(rounded_counts,
+                       housekeeping_genes[housekeeping_genes %in% rownames(seuratObj)], 
+                       k = k)
+  seuratObj[['RUVg']] <- Seurat::CreateAssayObject(as(RUVg$normalizedCounts, "dgCMatrix"))
```

Comment:
same comment as above about name of target and also whether it's actually a good idea to store into another assay